tower-http/src/services/fs/parse_range.rs

```diff
@@ -0,0 +1,309 @@
+                        }
+                        let start = file_size_bytes - 1 - end;
+                        ranges.push(RangeInclusive::new(start, file_size_bytes - 1));
+                        continue;
+                    }
+                    return ParsedRangeHeader::Malformed(format!(
+                        "Range: {} is not acceptable, end of range not parseable.",
+                        range
+                    ));
+                }
+                if let Ok(start) = start.parse::<u64>() {
+                    if end == "" {
+                        ranges.push(RangeInclusive::new(start, file_size_bytes - 1));
+                        continue;
+                    }
+                    if let Ok(end) = end.parse::<u64>() {
+                        if end >= file_size_bytes {
+                            return ParsedRangeHeader::Unsatisfiable(format!(
+                                "Range: {} is not satisfiable, end of range exceeds file boundary.",
+                                range
+                            ));
+                        }
+                        ranges.push(RangeInclusive::new(start, end));
+                        continue;
+                    }
+                    return ParsedRangeHeader::Malformed(format!(
+                        "Range: {} is not acceptable, end of range not parseable.",
+                        range
+                    ));
+                }
+                return ParsedRangeHeader::Malformed(format!(
+                    "Range: {} is not acceptable, start of range not parseable.",
+                    range
+                ));
+            }
+            return ParsedRangeHeader::Malformed(format!(
+                "Range: {} is not acceptable, range does not contain any dashes.",
+                range
+            ));
+        }
+    } else {
+        return ParsedRangeHeader::Malformed(format!(
+            "Range: {} is not acceptable, range does not start with 'bytes='",
+            range
+        ));
+    }
+    if ranges.is_empty() {
+        panic!("Programming error parsing range {}", range)
+    } else {
+        if ranges.len() == 1 {
+            ParsedRangeHeader::Range(ranges)
+        } else if !overlaps(&ranges) {
+            ParsedRangeHeader::Range(ranges)
+        } else {
+            return ParsedRangeHeader::Unsatisfiable(format!(
+                "Range header: {} is not satisfiable, ranges overlap",
+                range
+            ));
+        }
+    }
+}
+
+fn overlaps(ranges: &Vec<RangeInclusive<u64>>) -> bool {
+    let mut bounds = Vec::new();
+    for range in ranges {
+        bounds.push((range.start(), range.end()));
+    }
+    for i in 0..bounds.len() {
+        for j in i + 1..bounds.len() {
+            if bounds[i].0 <= bounds[j].1 && bounds[j].0 <= bounds[i].1 {
+                return true;
+            }
+        }
+    }
```

Comment:
I'm wondering if there is a way to write these loops without looping over the indexes 🤔

Something like:

```rust
for x in &thing {
    for y in &thing {
    }
}
```

That would also avoid potential panics from `bounds[i]` being out of bounds (no pun intended 😜)

Reply:
I was thinking about that but we shouldn't be able to go out of bounds here as `i` and `j` are both constricted, additionally the unconstrained double loob will make a comparison that is in the worst case n*n, while the constrained one will be a lot fewer, 
for non overlapping ranges a, b, and c we in that case get 
`aa, ab, ac,

ba, bb, bc,

ca, cb, cc`

vs

` ab, ac,

  bc`

Reply:
Hm 🤔 This code is only needed if the client sends multiple ranges right? If so, could the parsing return early and avoid these checks all together?

Reply:
If the length of ranges is 1 it doesn't check overlap